### PR TITLE
util-stats: Limit the max stats in InMemoryStatsReceiver

### PR DESCRIPTION
**Pull request by takezoe**
## Problem

`InMemoryStatsReceiver` keeps all added stats in memory so it consumes the heap infinitely.

## Solution

This pull request enables to limit the maximum number of stats stored in memory. The default is no limitation as before.

### CODE REVIEW

**Comment by mgodave** [comment]
@takezoe thanks for the contribution! Would it be possible to add some tests around the new functionality.

File: util-stats/src/test/java/com/twitter/finagle/stats/StatsReceiverCompilationTest.java [hunk 1]
```diff
@@ -23,7 +23,7 @@ public void testNullStatsReceiver() {
 
   @Test
   public void testStatMethods() {
-    InMemoryStatsReceiver sr = new InMemoryStatsReceiver();
+    InMemoryStatsReceiver sr = new InMemoryStatsReceiver(0);
```
Comment (mgodave) [inline util-stats/src/test/java/com/twitter/finagle/stats/StatsReceiverCompilationTest.java]: are these changes necessary since the default argument is zero?

File: util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala [hunk 2]
```diff
@@ -47,8 +47,10 @@ object InMemoryStatsReceiver {
  * assert(isr.counters(Seq("a", "b", "foo") == 0)) // ok
  * assert(isr.counters(Seq("a", "b", "bar") == 0)) // fail
  * }}}
+ *
+ * @param maxStats the maximum number of stats, zero or negative value means unlimited (default is unlimited)
  **/
-class InMemoryStatsReceiver extends StatsReceiver with WithHistogramDetails {
+class InMemoryStatsReceiver(maxStats: Int = 0) extends StatsReceiver with WithHistogramDetails {
```
Comment (mosesn) [inline util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala]: instead of using a default arg, please overload the constructor so it's java friendly (java doesn't know how to deal with default args).

also, would it more clearly communicate our intent to pass this as an Option, and require that it's > 0 if specified?

File: util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala [hunk 3]
```diff
@@ -114,7 +116,16 @@ class InMemoryStatsReceiver extends StatsReceiver with WithHistogramDetails {
 
       def add(value: Float): Unit = stats.synchronized {
         val oldValue = apply()
-        stats(schema.metricBuilder.name) = oldValue :+ value
+
+        // Limit the maximum number of stats as 10000000
```
Comment (mosesn) [inline util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala]: where does `10000000` come from?

File: util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala [hunk 4]
```diff
@@ -114,7 +116,16 @@ class InMemoryStatsReceiver extends StatsReceiver with WithHistogramDetails {
...
+        val length = oldValue.length
+        val tailValue = if (maxStats <= 0 || length < maxStats) {
```
Comment (mosesn) [inline util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala]: ffti: `stats(schema.metricBuilder.name) = oldValue.takeRight(maxStats) :+ value`

File: util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala (hunk 3 above)
Comment (takezoe) [reply util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala]: Oh, it came from my experiment code. Just removed thet comment.

File: util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala (hunk 4 above)
Comment (takezoe) [reply util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala]: Thanks! Updated.

File: util-stats/src/test/java/com/twitter/finagle/stats/StatsReceiverCompilationTest.java (hunk 1 above)
Comment (takezoe) [reply util-stats/src/test/java/com/twitter/finagle/stats/StatsReceiverCompilationTest.java]: They are necessary in Java code.

**Comment by takezoe** [comment]
Thanks for your review!

We are using our own in-memory stats receiver implementation on Finagle in production and we just thought we can use twitter-util's `InMemoryStatsReceiver` instead of ours if it supports expiration of recorded stats.

If this contradicts the design intention of `InMemoryStatsReceiver`, feel free to reject this pull request. We can understand your decision and we can continue to use our own in-memory solution.

File: util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala (hunk 2 above)
Comment (takezoe) [reply util-stats/src/main/scala/com/twitter/finagle/stats/InMemoryStatsReceiver.scala]: Yeah, so that we don't need to give a parameter in Java code.

**Comment by mosesn** [comment]
@takezoe how about using `finagle-stats` instead?  Do you need the precision that InMemorySR gives you?  Can you tell me more about what you're using it for?  We have a pretty decent production-friendly metrics implementation in `finagle-stats-core` that can extended for many other uses, or if you're just polling the metrics you can integrate with `TwitterServer` to export them from `finagle-stats` pretty automatically.

**Comment by mosesn** [comment]
Also, just fair warning–we might be a bit less responsive for the next couple of weeks, much of the team will be out for the holidays, but please keep commenting on the ticket and we'll try to get back to you when we can!

**Comment by takezoe** [comment]
Our use case is that exposing Finagle metrics via JMX. We have a common framework for exposing JMX metrics and currently, out JMX frameworks retrieves Finagle metrics through a modified version of `InMemoryStatsReceiver`.

I looked through finagle-stats-core roughly. Maybe we can use it instead.

**Comment by takezoe** [comment]
One thing I want to know about `MetricsStatsReceiver` which is provided by finagle-stats-core is how can we read metrics from `Metrics` although it's a package private.
https://github.com/twitter/finagle/blob/d42c87a97f25573759c1a6f605d0d2265f76619c/finagle-stats-core/src/main/scala/com/twitter/finagle/stats/Metrics.scala#L109

It might be a solution for JMX MBean to retrieves metrics from `/admin/metrics.json` which is served by `MetricsExporer`, but it's obviously not the best way. Rather, it might be better to implement our own receiver?

**Comment by mosesn** [comment]
@takezoe yeah, I think it might be good to change `MetricsStore` and `MetricsView` to be public, change Metrics.createDetached to return a `MetricsStore with MetricsView` and then you can use Metrics.createDetached to make a nice backend for exporting to JMX.  Alternatively, if you're using codahale's metrics, someone built a connector called [finagle-metrics](https://github.com/rlazoti/finagle-metrics) a few years ago.  It's not up to date, but you could probably fork it and dust it off if that happens to be your exact use case (I know they can export to JMX, so that's just a shot in the dark).

I think that will be substantially higher performance than InMemoryStatsReceiver, which is intentionally naive.

**Comment by takezoe** [comment]
Thanks! Since there are some better ways, I close this pull request and will try them.
